kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,207 @@
+
+## Design Details
+
+The design includes the following key components: the disk IO scheduler plugin and its interfaces with the IO Driver.
+
+<p align="center"><img src="images/key_components.png" title="Key components" width="600" class="center"/></p>
+The disk IO scheduler plugin communicates with the IO Driver to pass information on IO metrics collection context (e.g., reserved pod list), obtain a normalized IO BW for each new pod’s IO BW request, and retrieve updates on each disk’s normalized available IO capacity for making scheduling decisions.
+
+The IO Driver, which is to be implemented by disk IO vendors, comprises three components.
+
+### IO Metrics Collector
+
+The IO Metrics Collector, which runs on each worker node, acts as an IO metric collector and analyzer. It watches the actual disk IO utilization of each pod, calculates the disk’s available IO capacity based on each workload’s real-time usage and characteristics using the IO Calculation Model and reports the node’s real-time available disk IO capacity to the aggregator when it is smaller than what is saved in the scheduler plugin’s cache. Since different disk vendors could have different ways of collecting metric, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Aggregator
+
+The aggregator consolidates the IO metrics from each worker node, reports a consolidated list of real-time available disk IO capacity to the scheduler plugin. It is also responsible for converting each new pod’s disk IO BW request to a normalized value using the disk’s IO Calculation Model so as to match the disk’s normalized available IO capacity. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### IO Calculation Model
+
+The IO Calculation Model is responsible for converting the disk’s available IO capacity and each new pod’s IO request to normalized values. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Disk IO Scheduler Plugin
+
+We leverage the K8s scheduler framework to add the disk IO scheduler plugin.
+When the scheduler plugin starts, it subscribes to the IO Driver with a long streaming connection. The IO Driver updates the scheduler plugin about each node disk’s normalized available IO capacity and the scheduler plugin stores the info in the plugin’s local cache.
```

Comment:
Since the communication here done between the scheduler and a 3rd party component (and not the API server) are there any safety mechanism against timeout/slow responses, etc.?

Reply:
The safety mechanism is similar to scheduler extender. And the normalize function will be called when the pod has the disk IO specification. Moreover, unlike the scheduler extender, the IO driver will only be called once at the filter hook point.

Reply:
If I got @Tal-or right, I think he's referring to the fact that adhering closely to the standard kubernetes model, with node agents which push updates to objects (be them CRDs or native objects) to the apiserver, and with the client (scheduler) which watches these changes, brings for free the reliability and scalability guarantees ensured by the project framework. Sure, we need to adhere to the project's best practices, but once that is done the ecosystem basically solves that problem for free.
Diverging by that model means that to a possibly large extent the reimplementor has to take care of these concerns. This may be warranted by the specific case, but I think it should be clearly outlined why the k8s framework is not good enough and why changes to enable the specific use case are undesirable or even impossible.

Reply:
We thought about using CRD before. But going through the CRD route has two drawbacks: 1) In the pod scheduling flow, the disk IO scheduler plugin needs to pass the disk IO info specified in the annotation of the pod spec to the vendor-specific IO driver and get back a normalized disk IO BW value. Using CRD to support this bi-directional communication would involve creating two CRDs and two watches, which introduces long latency to the pod scheduling flow. 2) A node's disk IO available capacity would change dynamically upon the workloads' real-time IO access characteristics such as block size and IO read/write ratio. If we go through the CRD route, the real-time disk IO available capacity update will inject too much traffic into the API server and degrade the API server performance. Using a direct communication channel between the disk IO scheduler plugin and the vendor-specific IO driver greatly helps mitigate these two issues

Reply:
Thansk @cathyhongzhang for adding extra context, this helps. I recommend to add this information to the KEP text proper. I for myself I'll think a bit more about the problem and try to suggest alternatives. For alpha stage out-of-tree plugin I think the gRPC approach can work, taking into account the fact that scalability and communication issues would need to be pretty much reimplemented.  I'll defer to more experienced scheduler approver the final architectural review.

Reply:
@ffromani We will add this information to the KEP. Thanks for your review comments!